Refactor the visible_message() proc to be faster in some cases.

phil235 opened the pull request:
<s>It uses the player list instead of inspecting all atoms around the source (since visible actions don't need to be picked up by radios and other listening devices and clientless mobs). The proc should now be 2 to 6 times faster. It will only be slower than the current version if there is 400 players in a single room.</s>
The proc will no longer call get_hearers_in_view() (or equivalent) twice when the proc has a blind_message argument. This should make the proc faster in this specific case. The added client check will also make it faster when around many non player mobs.

Also:
- Replacing some "for(var/mob/O in viewers()) show_message()" by "visible_message()". This fixes pAI and mech occupant not getting these messages.
- Replacing some "show_message()" by " << "..." " instead. This removes unnecessary checks and prevents some edge cases like an admin not seeing a debug message because their mob is unconscious or blind.
- Fixes being able to see the results of the health and slime analyzers while blind.
- Removing some unnecessary "text(...)".

phil235 committed:
- Refactor the visible_message() proc to be much faster.
- Replacing some "for(var/mob/o in viewers()) show_message()" by "visible_message()".
- Replacing some "show_message()" by " << "..." " instead.
- Fixes being able to see the results of the health and slime analyzers while blind.
-

Bawhoppen commented:
we might as well call off the feature freeze

KorPhaeron commented:
Neat

optimumtact commented:
@Bawhoppen what?

Iamgoofball commented:
@Bawhoppen stop being a faggot, you know damn well that performance improvements are fine

Bawhoppen commented:
its not just /this/

phil235 commented:
@Bawhoppen There's no features in here, only fixes and a performance improvement.

Iamgoofball commented:
you aren't making any sense

are you talking about the other bugfix bundled in?

phil235 committed:
derp, some typos and a wrong number in the distance check.

Bawhoppen commented:
goof I'm talking about the other times the freeze was broken

phil235 commented on code/game/objects/items/devices/scanners.dm:
```diff
@@ -114,7 +114,8 @@ MASS SPECTROMETER
 /proc/healthscan(mob/living/user, mob/living/M, mode = 1)
-
+	if(user.stat || user.eye_blind)
+		return
```
I had fixed this bug once already but somebody reverted it somehow. You shouldn't see the analyzer's results if unconscious or blind.

phil235 commented on code/modules/admin/verbs/debug.dm:
```diff
@@ -201,7 +201,7 @@ But you can call procs that are of type /mob/living/carbon/human/proc/ for that
 	t+= "CO2: [env.carbon_dioxide]\n"
 
-	usr.show_message(t, 1)
+	usr << t
```
>Use debug proc, can't see the message because you're blind/unconscious, derp.

phil235 commented on code/modules/games/cards.dm:
```diff
@@ -185,10 +185,10 @@
 
 	if((!concealed || src.loc == usr) && cards.len)
-		usr.show_message("It contains: ")
+		usr.show_message("It contains: ", 1)
```
You shouldn't be able to see the card's contents if you're blind, thus the "1" argument.

phil235 commented on code/modules/mob/mob.dm:
```diff
@@ -42,108 +42,100 @@ var/next_mob_id = 0
+	usr << t
 
-/mob/proc/show_message(msg, type, alt, alt_type)//Message, type of message (1 or 2), alternative message, alt message type (1 or 2)
+/mob/proc/show_message(msg, type, alt_msg, alt_type)//Message, type of message (1 or 2), alternative message, alt message type (1 or 2)
```
changing arg name from "alt" to "alt_msg", because "alt" is a shitty arg name.

phil235 commented on code/modules/mob/mob.dm:
```diff
@@ -42,108 +42,100 @@ var/next_mob_id = 0
-		if(type & 1 && (disabilities & BLIND || paralysis) )//Vision related
-			if (!( alt ))
+	if(type)
+		if(type & 1 && eye_blind )//Vision related
```
eye_blind already covers all cases that "(disabilities & BLIND || paralysis)" and some other cases like being flashbanged or wearing welding helmet+sunglasses.

phil235 commented on code/modules/mob/mob.dm:
```diff
@@ -42,108 +42,100 @@ var/next_mob_id = 0
-	if(stat == UNCONSCIOUS || sleeping > 0)
-		src << "<I>... You can almost hear someone talking ...</I>"
+	// voice muffling
+	if(stat == UNCONSCIOUS)
```
you can't be sleeping but not unconscious.

phil235 commented on code/modules/mob/mob.dm:
```diff
@@ -42,108 +42,100 @@ var/next_mob_id = 0
+	// voice muffling
+	if(stat == UNCONSCIOUS)
+		if(type & 2) //audio
+			src << "<I>... You can almost hear something ...</I>"
```
I changed the wording a bit because this does not cover only say speech. The check is here so you don't get that message when unconscious and a blob attacks you or you're hit by a floral somatoray beam for example.

duncathan commented:
@Bawhoppen what the fuck are you talking about; has the freeze even been broken once?

phil235 committed:
- switch visible_message() back to using get_hearers_in_view(), because using inLineOfSight() creates some bugs. visible_message() will still be faster than current version, but only when there's a blind_message used.
- Fixes hostile animal not seeing any target near them if they are sitting on something opaque (wall, mech,..).

phil235 commented:
Using inLineOfSight() was causing some issue, I had to go back to using get_hearers_in_view() :disappointed:

duncathan commented on code/modules/mob/mob.dm:
```diff
@@ -42,108 +42,94 @@ var/next_mob_id = 0
-		if (type & 2 && ear_deaf)//Hearing related
-			if (!( alt ))
+
+		if(type & 2 && ear_deaf)//Hearing related
```
Defines please

duncathan replied:
@phil235

phil235 replied:
will fix when I have the time.

Cheridan commented:
This is good enough to merge now, the defines can be added later
